### PR TITLE
Add each PropertySource from PropertySourceLocators individually

### LINKED ISSUE #611: Replace CompositePropertySource with one PropertySource per file
See https://github.com/spring-projects/spring-boot/issues/18174 for background and a sample application.

Raising an issue here as requested by @spencergibb

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Fixes #611

### CODE REVIEW

File: spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java [hunk 1]
```diff
@@ -154,36 +164,39 @@ private void setLogLevels(ConfigurableApplicationContext applicationContext,
...
+			List<PropertySource<?>> composite) {
 		MutablePropertySources incoming = new MutablePropertySources();
-		incoming.addFirst(composite);
+		List<PropertySource<?>> reversedComposite = new ArrayList(composite);
+		// Reverse the list so that when we call addFirst below we are maintaining the
+		// same order of PropertySournces
+		// Wherever we call addLast we can use the order in the List since the first item
+		// will end up before the rest
+		Collections.reverse(reversedComposite);
+		reversedComposite.stream().forEach(p -> incoming.addFirst(p));
```
Comment (spencergibb) [inline spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: This line seems out of place to me.

File: spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java [hunk 2]
```diff
@@ -241,3 +254,32 @@ private void handleIncludedProfiles(ConfigurableEnvironment environment) {
...
+	@Override
+	public Object getProperty(String name) {
+		return p.getProperty(name);
+	}
+
+	@Override
+	public String[] getPropertyNames() {
+		Set<String> names = new LinkedHashSet<>();
+		if (!(this.p instanceof EnumerablePropertySource)) {
+			throw new IllegalStateException(
```
Comment (spencergibb) [inline spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: Should it fail or log a warning?

File: spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java (hunk 1 above)
Comment (ryanjbaxter) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: I could move `Collections.reverse(reversedComposite);` to the line above so it is a bit clearer.  
```
List<PropertySource<?>> reversedComposite = Collections.reverse(new ArrayList(composite));
```

As you know, the property source list matters....

Consider we have a `composite` of PropertySource1, PropertySource2, and PropertySource3.  If I just iterated through the list and called `incoming.addFirst(p)` the resulting `incoming` property source list would have PropertySource3, PropertySource2, PropertySource1, essentially the reverse order.  So I reverse the list in order to make sure the property sources in `composite` have the same order.

File: spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java (hunk 2 above)
Comment (ryanjbaxter) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: I took this code from the implementation in `CompositePropertySource` it just throws the exception, so figured I would do the same

File: spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java (hunk 1 above)
Comment (spencergibb) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: :+1:

File: spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java [hunk 3]
```diff
@@ -154,36 +164,39 @@ private void setLogLevels(ConfigurableApplicationContext applicationContext,
...
+		// will end up before the rest
+		Collections.reverse(reversedComposite);
+		reversedComposite.stream().forEach(p -> incoming.addFirst(p));
 		PropertySourceBootstrapProperties remoteProperties = new PropertySourceBootstrapProperties();
 		Binder.get(environment(incoming)).bind("spring.cloud.config",
 				Bindable.ofInstance(remoteProperties));
 		if (!remoteProperties.isAllowOverride() || (!remoteProperties.isOverrideNone()
 				&& remoteProperties.isOverrideSystemProperties())) {
-			propertySources.addFirst(composite);
+			reversedComposite.stream().forEach(p -> propertySources.addFirst(p));
```
Comment (dsyer) [inline spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: I'm a bit uncomfortable with using Stream APIs in this low level code. Startup and GC performance is quite sensitive to this kind of thing. Can we use a normal iterator?

Comment (spencergibb) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: :+1:

Comment (ryanjbaxter) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: For sure

Comment (ryanjbaxter) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: @dsyer done!

Comment (spencergibb) [reply spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: :+1:

**Comment by dsyer** [comment]
LGTM

**Pull request merged**